File: examples/local-triage/renderers/performance/native.ts
/** Exact packaged CLI/TUI/desktop qualification. Run through mac-native custody. */
import { mkdir, readFile, writeFile } from "node:fs/promises";
import { join, resolve } from "node:path";
import { createHash } from "node:crypto";
import budgets from "./budgets.json";

const packageRoot = process.argv[2], output = process.argv[3];
if (!packageRoot || !output || process.platform !== "darwin") throw new Error("Usage on macOS: bun native.ts PACKAGE_ROOT FRESH_OUTPUT");
const directory = resolve(output), pkg = resolve(packageRoot), state = join(directory, "state");
await mkdir(directory, { recursive: true });
const hash = (bytes: Uint8Array | string) => createHash("sha256").update(bytes).digest("hex");
const manifest = JSON.parse(await readFile(join(pkg, "manifest.json"), "utf8")) as { provenance: { gitHead: string; source: { sha256: string } }; files: { path: string; sha256: string }[] };
const declared = { contract: "algal.renderer-performance-declaration.v1", declaredAt: new Date().toISOString(), budgets, budgetsSha256: hash(await readFile(new URL("./budgets.json", import.meta.url))), scriptSha256: hash(await readFile(import.meta.path)), helperSha256: hash(await readFile(new URL("./native.swift", import.meta.url))), packageSource: manifest.provenance.source.sha256, gitHead: manifest.provenance.gitHead };
await writeFile(join(directory, "declaration.json"), JSON.stringify(declared, null, 2) + "\n", { flag: "wx" });
const checks: { name: string; passed: boolean; observed: unknown; maximum?: number }[] = [];
const check = (name: string, passed: boolean, observed: unknown, maximum?: number) => checks.push({ name, passed, observed, ...(maximum === undefined ? {} : { maximum }) });
const quantiles = (values: number[]) => ({ count: values.length, min: Math.min(...values), median: [...values].sort((a, b) => a - b)[Math.floor(values.length / 2)]!, p95: [...values].sort((a, b) => a - b)[Math.ceil(values.length * 0.95) - 1]!, max: Math.max(...values) });
async function command(args: string[], deadlineMs = 30000) {
  const started = performance.now(), child = Bun.spawn(args, { cwd: directory, stdin: "ignore", stdout: "pipe", stderr: "pipe" });
  const deadline = setTimeout(() => child.kill("SIGKILL"), deadlineMs);
  try {
    const [stdout, stderr, status] = await Promise.all([new Response(child.stdout).text(), new Response(child.stderr).text(), child.exited]);
    if (stdout.length > 524288 || stderr.length > 16384) throw new Error("Qualification output exceeded its bound");
    return { stdout, stderr, status, ms: performance.now() - started };
  } finally { clearTimeout(deadline); if (child.exitCode === null) child.kill("SIGKILL"); await child.exited; }
}
for (const path of ["bin/triage-host", "bin/triage-tui", "ALGAL Triage.app/Contents/MacOS/triage-desktop"]) {
  const expected = manifest.files.find(file => file.path === path);
  if (!expected || hash(await readFile(join(pkg, path))) !== expected.sha256) throw new Error("Package identity mismatch");
}
const seed = { schemaVersion: 2, config: { sort: "priority", group: "status", allowReopen: true }, tasks: Array.from({ length: budgets.local.tasks }, (_, i) => ({ id: `task-${i}`, title: `Performance task ${String(i).padStart(2, "0")}`, priority: "normal", status: "open", category: "inbox" })) };
const seedPath = join(directory, "seed.json"); await writeFile(seedPath, JSON.stringify(seed));
const init = await command([join(pkg, "bin/triage-host"), state, "performance", "init", seedPath]);
if (init.status !== 0) throw new Error(init.stderr);
const hostSamples: number[] = [], terminalSamples: number[] = [], peakRss: number[] = [];
for (const [samples, args] of [
  [hostSamples, [join(pkg, "bin/triage-host"), state, "performance", "capture"]],
  [terminalSamples, [join(pkg, "bin/triage-tui"), "--dir", state, "--application", "performance", "--snapshot"]],
] as const) {
  for (let i = 0; i < budgets.local.samples; i++) {
    const result = await command(["/usr/bin/time", "-l", ...args]);
    if (result.status !== 0) throw new Error(result.stderr);
    samples.push(result.ms);
    const rss = /([0-9]+)\s+maximum resident set size/.exec(result.stderr);
    if (!rss) throw new Error("Darwin time did not return peak resident memory");
    peakRss.push(Number(rss[1]));
  }
}
const host = quantiles(hostSamples), terminal = quantiles(terminalSamples);
check("Packaged host cold process capture p95", host.p95 <= budgets.local.hostCaptureP95MsMax, host, budgets.local.hostCaptureP95MsMax);
check("Packaged TUI process-to-complete-snapshot p95", terminal.p95 <= budgets.local.terminalStartupP95MsMax, terminal, budgets.local.terminalStartupP95MsMax);
check("CLI/TUI Darwin peak resident bytes", Math.max(...peakRss) <= budgets.local.processPeakRssBytesMax, peakRss, budgets.local.processPeakRssBytesMax);
const helper = join(directory, "native-observer");
const compiled = await command(["/usr/bin/xcrun", "swiftc", new URL("./native.swift", import.meta.url).pathname, "-o", helper], 60000);
if (compiled.status !== 0) throw new Error(compiled.stderr);
const started = Date.now(), desktop = Bun.spawn([join(pkg, "ALGAL Triage.app/Contents/MacOS/triage-desktop"), "--dir", state, "--application", "performance"], { cwd: directory, stdin: "ignore", stdout: "ignore", stderr: "ignore" });
let native: { ok: boolean; available: boolean; reason?: string; readyMs?: number; actionsMs?: number[]; mainProcessRssBeforeBytes?: number; mainProcessRssAfterBytes?: number };
try {
  const observed = await command([helper, String(desktop.pid), String(started), String(budgets.local.desktopActions)], 210000);
  native = JSON.parse(observed.stdout);
} finally { if (desktop.exitCode === null) desktop.kill("SIGTERM"); await desktop.exited; }
check("Desktop content becomes accessibility-visible", native.ok && native.readyMs !== undefined && native.readyMs <= budgets.local.desktopReadyMsMax, native.readyMs ?? native.reason, budgets.local.desktopReadyMsMax);
const actions = native.actionsMs?.length ? quantiles(native.actionsMs) : null;
check("Desktop complete/reopen action p95 including AX observation", native.ok && actions !== null && actions.p95 <= budgets.local.desktopActionP95MsMax, actions, budgets.local.desktopActionP95MsMax);
const growth = native.mainProcessRssAfterBytes !== undefined && native.mainProcessRssBeforeBytes !== undefined ? native.mainProcessRssAfterBytes - native.mainProcessRssBeforeBytes : null;
check("Desktop main-process retained growth after task updates", native.ok && growth !== null && growth <= budgets.local.desktopMainRssGrowthBytesMax, growth, budgets.local.desktopMainRssGrowthBytesMax);
const report = { contract: "algal.renderer-native-performance.v1", ok: checks.every(item => item.passed), declaration: declared, platform: { platform: process.platform, arch: process.arch }, initMs: init.ms, host, terminal, native, checks, limitations: ["Local development Mac and bounded 32-task profile only; not a production SLO or cross-platform benchmark", "CLI/TUI memory is Darwin time maximum resident size; desktop growth covers its main process only, excluding separately managed WebKit processes", "Desktop action latency includes bounded accessibility-tree observation overhead", "No model call, provider network operation or generated revision occurred in this measurement"] };
await writeFile(join(directory, "result.json"), JSON.stringify(report, null, 2) + "\n", { flag: "wx" });
console.log(JSON.stringify({ ok: report.ok, report: join(directory, "result.json"), checks }));
if (!report.ok) process.exitCode = 1;

File: examples/local-triage/renderers/performance/native.swift
import AppKit
import ApplicationServices
import Foundation

// Read only this task-owned application's accessibility tree. Never request a
// new permission, inspect another app, or use inference to perform the actions.
func attribute(_ node: AXUIElement, _ name: CFString) -> CFTypeRef? {
    var value: CFTypeRef?
    return AXUIElementCopyAttributeValue(node, name, &value) == .success ? value : nil
}
func nodes(_ root: AXUIElement) -> [AXUIElement] {
    var result: [AXUIElement] = []
    var queue: [(AXUIElement, Int)] = [(root, 0)]
    var cursor = 0
    while cursor < queue.count && cursor < 4096 {
        let (node, depth) = queue[cursor]; cursor += 1; result.append(node)
        if depth < 24, let children = attribute(node, kAXChildrenAttribute as CFString) as? [AXUIElement] {
            for child in children.prefix(256) { queue.append((child, depth + 1)) }
        }
    }
    return result
}
func text(_ node: AXUIElement) -> [String] {
    [kAXTitleAttribute, kAXValueAttribute, kAXDescriptionAttribute].compactMap {
        attribute(node, $0 as CFString) as? String
    }
}
func button(_ nodes: [AXUIElement], _ label: String) -> AXUIElement? {
    nodes.first { (attribute($0, kAXRoleAttribute as CFString) as? String) == kAXButtonRole && text($0).contains(label) }
}
func waitFor(_ predicate: () -> Bool, seconds: Double = 8) -> Bool {
    let deadline = ProcessInfo.processInfo.systemUptime + seconds
    repeat { if predicate() { return true }; Thread.sleep(forTimeInterval: 0.02) }
    while ProcessInfo.processInfo.systemUptime < deadline
    return false
}
func rss(_ pid: Int32) -> Int {
    let process = Process(); process.executableURL = URL(fileURLWithPath: "/bin/ps")
    process.arguments = ["-o", "rss=", "-p", String(pid)]
    let pipe = Pipe(); process.standardOutput = pipe; process.standardError = FileHandle.nullDevice
    do { try process.run(); let bytes = pipe.fileHandleForReading.readDataToEndOfFile(); process.waitUntilExit()
        return (Int(String(decoding: bytes, as: UTF8.self).trimmingCharacters(in: .whitespacesAndNewlines)) ?? 0) * 1024
    } catch { return 0 }
}
func emit(_ object: [String: Any], code: Int32) -> Never {
    if let bytes = try? JSONSerialization.data(withJSONObject: object, options: [.sortedKeys]) {
        print(String(decoding: bytes, as: UTF8.self))
    }
    exit(code)
}
guard CommandLine.arguments.count == 4,
      let pid = Int32(CommandLine.arguments[1]),
      let started = Double(CommandLine.arguments[2]),
      let count = Int(CommandLine.arguments[3]), count > 0 && count <= 24 else {
    emit(["ok": false, "reason": "Expected PID, start epoch milliseconds and at most 24 actions"], code: 2)
}
guard AXIsProcessTrusted() else {
    emit(["ok": false, "available": false, "reason": "Inherited accessibility permission unavailable; no permission prompt was requested"], code: 2)
}
let app = AXUIElementCreateApplication(pid)
guard waitFor({ nodes(app).contains { text($0).contains { $0.contains("Performance task") } } }, seconds: 12) else {
    emit(["ok": false, "available": true, "reason": "Task content did not become accessibility-visible within the readiness deadline"], code: 3)
}
let readyMs = Date().timeIntervalSince1970 * 1000 - started
let beforeRss = rss(pid)
var elapsed: [Double] = []
for index in 0..<count {
    let completing = index % 2 == 0
    guard let target = button(nodes(app), completing ? "Complete" : "Reopen") else {
        emit(["ok": false, "available": true, "reason": "Expected captured task action unavailable", "completed": index], code: 4)
    }
    let begin = ProcessInfo.processInfo.systemUptime
    guard AXUIElementPerformAction(target, kAXPressAction as CFString) == .success,
          waitFor({ (button(nodes(app), "Reopen") != nil) == completing }) else {
        emit(["ok": false, "available": true, "reason": "Task action did not reach its displayed result", "completed": index], code: 5)
    }
    elapsed.append((ProcessInfo.processInfo.systemUptime - begin) * 1000)
}
Thread.sleep(forTimeInterval: 0.2)
emit(["ok": true, "available": true, "readyMs": readyMs, "actionsMs": elapsed,
      "mainProcessRssBeforeBytes": beforeRss, "mainProcessRssAfterBytes": rss(pid),
      "memoryScope": "Desktop main process only; separately managed WebKit processes are excluded",
      "timingScope": "Application launch to accessibility-visible task content; action includes AX observation overhead"], code: 0)
